refactoring of print table to not overwrite borders with background color

stkruse opened the pull request:
This is an implementation of wxPdfTable::Write() which first finds page breaks, than draws on every page at first the background, than the border and at last the content.
So, the cell border is not overwritten by the background of a subsequent table cell.

The old WriteRow() is not used anymore but still available since it is part of the public interface of the class.

In wxPdfDocument a member has been added to save the height of the header on each page. This value is needed to calculate the page breaks.

utelle commented on src/pdfform.cpp:
```diff
@@ -655,3 +655,8 @@ wxPdfDocument::LoadZapfDingBats()
   }
 }
+
+double wxPdfDocument::GetHeaderHeight() const
```
The implementation of `wxPdfDocument::GetHeaderHeight()` has nothing to do with PDF forms. Therefore it should be moved to `pdfdocument.cpp`. (Since the method is protected, another possibility would be to move it to `pdfkernel.cpp`, but I think `pdfdocument.cpp` is more suitable, as we may consider to make this method public at a later time.)

stkruse replied:
Yes, you're absolutely correct. Somehow the C++ Resharper played me a prank. I declared it in ```pdfdocument.h``` and just clicked on "create definition" and didn't reviewed the source file. I assumed it is in ```pdfdocument.cpp```. I'll adjust it, but probably not today.

utelle replied:
Thanks in advance. (It's not urgent ...)

utelle commented on src/pdfxml.cpp:
```diff
@@ -508,58 +508,363 @@ wxPdfTable::InsertCell(wxPdfTableCell* cell)
+  const wxArrayInt firstRowsOnNextPage = GetFirstRowsOnNextPage();
+
+  //Case we get everything on the same Page
+  if(firstRowsOnNextPage.empty())
```
Maybe it is not necessary to separately handle the case where the whole table fits on one page, if you add a row number 1 larger than `m_bodyRowLast` as a sentinel to `firstRowsOnNextPage`. This would require only a small change to method `wxPdfTable::WriteOnMultiplePages`, namely checking whether the first entry is not the sentinel, before issuing a page break.

stkruse replied:
yes, there are different ways to solve it. I thought it could be more readable in that way than to enter a sentinel. How ever, somewhere it has to be checked if it's all on one page, even if it's done in ```GetFirstRowsOnNextPage```.

Another possibility would be to call ```WriteRowsOnPage(unsigned firstRow, unsigned lastRow, double x, double y, bool writeHeader)``` directly in the cited ```if```, which would make ```WriteOnPage``` obsolete.

utelle replied:
> yes, there are different ways to solve it. I thought it could be more readable in that way than to enter a sentinel. How ever, somewhere it has to be checked if it's all on one page, even if it's done in `GetFirstRowsOnNextPage`.

Of course, it has to be checked somewhere one way or the other, but using a sentinel entry in `firstRowsOnNextPage` makes this easy and eliminates any special handling.

> Another possibility would be to call `WriteRowsOnPage(unsigned firstRow, unsigned lastRow, double x, double y, bool writeHeader)` directly in the cited `if`, which would make `WriteOnPage` obsolete.

Well, actually I would prefer the solution with the sentinel entry. The only check that has to be made in `WriteOnMultiplePages` is when the sentinel is reached, so that not too many page breaks are issued. A method name like `WriteTable`, `WriteAll`, or `WriteAllRows` might be more appropriate than `WriteOnMultiplePages`, when using this approach.

utelle reviewed (changes requested):
Please check my comments.

Additionally, it would be nice if you could enhance the markup sample in the minimal sample. There is already a table that spans more than one page. Adding an additional column and maybe cell backgrounds would be nice. However, I can add that later myself.

utelle commented:
> The old WriteRow() is not used anymore but still available since it is part of the public interface of the class.

Yes, it is part of the public interface, but the whole class `wxPdfTable` is meant for internal use only. Therefore, I think method `WriteRow()` can be safely removed, if it is not used anymore. To keep the method would only make sense, if you intend to support non-collapsed borders based on the modifications you mentioned in #81.

stkruse commented:
Hi, 
I added three commits, so you can pick the commits you want to include:
First of all: move implementation of ```GetHeaderHeight()``` to ```pdfdocumen.cpp```. 
Than, I removed the now unused ```WriteRow```. 
And another approach for writing tables. Not via the first line on a possible new page, but via the last line on a page (+1 for iteration ;) )
It looks clearer and was also my first approach. To be honest, I don't remember why I dismissed this approach in the first time.

utelle commented:
> Hi, I added three commits, so you can pick the commits you want to include: First of all: move implementation of `GetHeaderHeight()` to `pdfdocumen.cpp`. Than, I removed the now unused `WriteRow`. And the third one - another approach for writing tables. Not via the first line on a possible new page, but via the last line on a page (+1 for iteration ;) ) It looks clearer and was also my first approach. To be honest, I don't remember why I dismissed this approach in the first time.

Thanks a lot for your efforts. I'm going to merge your modifications. However, I wait for the CI runs to complete - last time the macOS CI run failed due to a problem with setting up the environment.

utelle commented:
Well, the Ubuntu CI run fails in the testing step. The minimal sample crashes with a segmentation fault. In my local environment, I see the crash happen on executing the markup sample in method `wxPdfTable::WriteFillingOfRows` on accessing the row heights (`y += m_rowHeights.find(row)->second;`). Could you please check? Most likely, it will be only a minor glitch somewhere.

stkruse commented:
> Well, the Ubuntu CI run fails in the testing step. The minimal sample crashes with a segmentation fault. In my local environment, I see the crash happen on executing the markup sample in method `wxPdfTable::WriteFillingOfRows` on accessing the row heights (`y += m_rowHeights.find(row)->second;`). Could you please check? Most likely, it will be only a minor glitch somewhere.

Done... somehow I forgot to include my last local commit... should now be there.

utelle reviewed (approved):
Thanks for improving the handling of table cell borders.
